src/TestExplorer/LSPTestDiscovery.ts

```diff
@@ -29,104 +34,118 @@ import { LanguageClientManager } from "../sourcekit-lsp/LanguageClientManager";
+            if (textDocumentCaps?.version >= 2) {
+                const testsInDocument = await client.sendRequest(
+                    textDocumentTestsRequest,
+                    { textDocument: { uri: document.toString() } },
+                    token
+                );
+                return this.transform(client, swiftPackage, testsInDocument);
+            } else {
+                throw new Error("textDocument/tests requests not supported");
+            }
+        });
     }
 
     /**
      * Return list of workspace tests
      * @param workspaceRoot Root of current workspace folder
      */
-    async getWorkspaceTests(workspaceRoot: vscode.Uri): Promise<TestDiscovery.TestClass[]> {
+    async getWorkspaceTests(
+        swiftPackage: SwiftPackage,
+        workspaceRoot: vscode.Uri
+    ): Promise<TestDiscovery.TestClass[]> {
         return await this.languageClient.useLanguageClient(async (client, token) => {
-            const tests = await client.sendRequest(workspaceTestsRequest, {}, token);
-            const testsInWorkspace = tests.filter(item =>
-                isPathInsidePath(
-                    client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                    workspaceRoot.fsPath
-                )
-            );
-            const classes = testsInWorkspace
-                .filter(item => {
-                    return (
-                        item.kind === langclient.SymbolKind.Class &&
-                        isPathInsidePath(
-                            client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
-                            workspaceRoot.fsPath
-                        )
-                    );
-                })
-                .map(item => {
-                    const functions = testsInWorkspace
-                        .filter(func => func.containerName === item.name)
-                        .map(func => {
-                            const openBrackets = func.name.indexOf("(");
-                            let funcName = func.name;
-                            if (openBrackets) {
-                                funcName = func.name.slice(0, openBrackets);
-                            }
-                            return {
-                                name: funcName,
-                                location: client.protocol2CodeConverter.asLocation(func.location),
-                            };
-                        });
-                    return {
-                        name: item.name,
-                        location: client.protocol2CodeConverter.asLocation(item.location),
-                        functions: functions,
-                    };
-                });
-            return classes;
+            const workspaceTestCaps = client.initializeResult?.capabilities.experimental;
+            if (!workspaceTestCaps) {
+                throw new Error("textDocument/tests requests not supported");
```

Comment:
todo: "workspace/tests"

suggestion: Since same logic for textDocument, we could extract this into it's own function so we can re-use if more experimental capabilities are added

Reply:
Good call, I've refactored out the common capability check into a method.